presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,283 @@
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.EQUAL;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class ReorderJoins
+        implements Rule
+{
+    private final CostComparator costComparator;
+
+    public ReorderJoins(CostComparator costComparator)
+    {
+        this.costComparator = costComparator;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        // We check that join distribution type is absent because we only want to do this transformation once (reordered joins will have distribution type already set).
+        if (!(node instanceof JoinNode) || !isJoinReorderingEnabled(session)) {
+            return Optional.empty();
+        }
+
+        JoinNode joinNode = (JoinNode) node;
+        if (!(joinNode.getType() == INNER) || joinNode.getDistributionType().isPresent()) {
+            return Optional.empty();
+        }
+
+        JoinGraphNode joinGraph = new JoinGraphNode.JoinGraphNodeBuilder(joinNode, lookup).toJoinGraphNode(idAllocator);
+        return Optional.of(new JoinEnumerator(idAllocator, symbolAllocator, session, lookup, joinGraph, costComparator).chooseJoinOrder(joinGraph));
+    }
+
+    @VisibleForTesting
+    public static class JoinEnumerator
+    {
+        private final Map<JoinGraphNode, JoinNode> memo = new HashMap<>();
+        private final PlanNodeIdAllocator idAllocator;
+        private final Session session;
+        private final Ordering<PlanNode> planNodeOrdering;
+        private final CostComparator costComparator;
+
+        public JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup, JoinGraphNode joinGraphNode, CostComparator costComparator)
+        {
+            this.idAllocator = idAllocator;
+            this.session = session;
+            this.planNodeOrdering = new Ordering<PlanNode>()
+            {
+                @Override
+                public int compare(PlanNode node1, PlanNode node2)
+                {
+                    PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return JoinEnumerator.this.costComparator.compare(session, node1Cost, node2Cost);
```

Comment:
`s/JoinEnumerator.this.//` for readability

then you can remove `costComparator` field